src/TokenUriResolver.sol

```diff
@@ -0,0 +1,113 @@
+        IJBTokenUriResolver indexed tokenUriResolver
+    );
+    event ProjectTokenUriResolverSet(
+        uint256 indexed projectId,
+        IJBTokenUriResolver indexed tokenUriResolver
+    );
+
+    /**
+     * @notice Each project's IJBTokenUriResolver metadata contract.
+     * @dev Mapping of projectId => tokenUriResolver
+     * @dev The 0 project returns the default resolver's address.
+     * @return IJBTokenUriResolver The address of the IJBTokenUriResolver for the project, or 0 if none is set.
+     */
+    mapping(uint256 => IJBTokenUriResolver) public tokenUriResolvers;
+
+    /**
+     * @notice TokenUriResolver constructor.
+     * @dev Sets the default IJBTokenUriResolver. This resolver is used for all projects that do not have a custom resolver.
+     * @param _projects The address of the Juicebox Projects contract.
+     * @param _operatorStore The address of the JBOperatorStore contract.
+     * @param _defaultTokenUriResolver The address of the default IJBTokenUriResolver.
+     */
+    constructor(
+        IJBProjects _projects,
+        IJBOperatorStore _operatorStore,
+        IJBTokenUriResolver _defaultTokenUriResolver
+    ) JBOperatable(_operatorStore) {
+        projects = _projects;
+        tokenUriResolvers[0] = IJBTokenUriResolver(_defaultTokenUriResolver);
+    }
+
+    /**
+     *  @notice Get the token uri for a project.
+     *  @dev Called by `JBProjects.tokenUri(uint256)`. If a project has a custom IJBTokenUriResolver, it is used instead of the default resolver.
+     *  @param _projectId The id of the project.
+     *  @return tokenUri The token uri for the project.
+     *  @inheritdoc	IJBTokenUriResolver
+     */
+    function getUri(uint256 _projectId)
+        external
+        view
+        override
+        returns (string memory tokenUri)
+    {
+        IJBTokenUriResolver tur = tokenUriResolvers[_projectId];
+        if (tur == IJBTokenUriResolver(address(0))) {
+            return tokenUriResolvers[0].getUri(_projectId);
+        } else {
+            try tur.getUri(_projectId) returns (string memory uri) {
+                return uri;
+            } catch {
+                return tokenUriResolvers[0].getUri(_projectId);
+            }
+        }
+    }
+
+    /**
+     * @notice Set the IJBTokenUriResolver for a project. This function is restricted to the project's owner and operators.
+     * @dev Set the IJBTokenUriResolver for a project to 0 to use the default resolver.
+     * @param _projectId The id of the project.
+     * @param _resolver The address of the IJBTokenUriResolver, or 0 to restore the default setting.
+     */
+    function setTokenUriResolverForProject(
+        uint256 _projectId,
+        IJBTokenUriResolver _resolver
+    )
+        external
+        requirePermission(
+            projects.ownerOf(_projectId),
+            _projectId,
+            JBUriOperations.SET_TOKEN_URI
+        )
+    {
+        if (_projectId == 0) revert UNAUTHORIZED();
```

Comment:
Is the requirePermission not reverting in this case?

Reply:
I *think* you're right. I pass the `projectId` in the `_domain` field of `requirePermission`, which I guess will fail in `return (((permissionsOf[_operator][_account][_domain] >> _permissionIndex) & 1) == 1)`? I'm not confident about how Operator Store works but I'll remove as your suggestion makes sense to me. https://docs.juicebox.money/dev/api/contracts/jboperatorstore/read/haspermission/#body

Reply:
Removed [824ffb0](https://github.com/jbx-protocol/juice-token-resolver/pull/1/commits/824ffb051283050422f2047861b22599f5369ebc)